chain/index/read.go

```diff
@@ -0,0 +1,135 @@
+	}
+
+	var msgCidBytes []byte
+
+	if err := si.readWithHeadIndexWait(ctx, func() error {
+		return si.queryMsgCidFromEthHash(ctx, txHash, &msgCidBytes)
+	}); err != nil {
+		return cid.Undef, err
+	}
+
+	msgCid, err := cid.Cast(msgCidBytes)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to cast message CID: %w", err)
+	}
+
+	return msgCid, nil
+}
+
+func (si *SqliteIndexer) queryMsgCidFromEthHash(ctx context.Context, txHash ethtypes.EthHash, msgCidBytes *[]byte) error {
+	return si.stmts.getMsgCidFromEthHashStmt.QueryRowContext(ctx, txHash.String()).Scan(msgCidBytes)
+}
+
+func (si *SqliteIndexer) GetMsgInfo(ctx context.Context, messageCid cid.Cid) (*MsgInfo, error) {
+	if si.isClosed() {
+		return nil, ErrClosed
+	}
+
+	var tipsetKeyCidBytes []byte
+	var height int64
+
+	if err := si.readWithHeadIndexWait(ctx, func() error {
+		return si.queryMsgInfo(ctx, messageCid, &tipsetKeyCidBytes, &height)
+	}); err != nil {
+		return nil, err
+	}
+
+	tipsetKey, err := cid.Cast(tipsetKeyCidBytes)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to cast tipset key cid: %w", err)
+	}
+
+	return &MsgInfo{
+		Message: messageCid,
+		TipSet:  tipsetKey,
+		Epoch:   abi.ChainEpoch(height),
+	}, nil
+}
+
+// This function attempts to read data using the provided readFunc.
+// If the initial read returns no rows, it waits for the head to be indexed
+// and tries again. This ensures that the most up-to-date data is checked.
+// If no data is found after the second attempt, it returns ErrNotFound.
+func (si *SqliteIndexer) readWithHeadIndexWait(ctx context.Context, readFunc func() error) error {
+	err := readFunc()
+	if err == sql.ErrNoRows {
+		// not found, but may be in latest head, so wait for it and check again
+		if err := si.waitTillHeadIndexed(ctx); err != nil {
+			return xerrors.Errorf("failed while waiting for head to be indexed: %w", err)
+		}
+		err = readFunc()
+	}
+
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return ErrNotFound
+		}
+		return xerrors.Errorf("failed to read data from index: %w", err)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) queryMsgInfo(ctx context.Context, messageCid cid.Cid, tipsetKeyCidBytes *[]byte, height *int64) error {
+	return si.stmts.getNonRevertedMsgInfoStmt.QueryRowContext(ctx, messageCid.Bytes()).Scan(tipsetKeyCidBytes, height)
+}
+
+func (si *SqliteIndexer) waitTillHeadIndexed(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, headIndexedWaitTimeout)
+	defer cancel()
+
+	head := si.cs.GetHeaviestTipSet()
+	headTsKeyCidBytes, err := toTipsetKeyCidBytes(head)
+	if err != nil {
+		return xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// wait till it is indexed
+	subCh, unsubFn := si.subscribeUpdates()
+	defer unsubFn()
+
+	for ctx.Err() == nil {
+		exists, err := si.isTipsetIndexed(ctx, headTsKeyCidBytes)
+		if err != nil {
+			return xerrors.Errorf("failed to check if tipset exists: %w", err)
+		} else if exists {
+			return nil
+		}
+
+		select {
```

Comment:
Can we add a long timeout in here, I'm a bit worried there might be some permutation of settings whereby you can be querying but also not properly indexing, or getting the heaviest tipset and by the time you ask for it there's a reason it's not indexes -- I can't see a way that any of that could happen, but this being an indefinite loop is a worry. How about a 90 - 120s timeout on this before it errors?

Reply:
Done.

Reply:
Done.